File: packages/forma-python/src/forma/agent.py
import hashlib
import json
from pathlib import Path
from typing import Any

from .provider import ModelProvider, provider_from_profile, provider_profile_from_file
from .runtime import FormaRuntime
from .types import FormaResult, FormaValue


class FormaAgent:
    def __init__(
        self,
        *,
        task: str,
        provider: ModelProvider | None = None,
        tools: dict[str, object] | None = None,
        source: str | None = None,
        source_name: str | None = None,
        file: str | Path | None = None,
    ) -> None:
        if (source is None) == (file is None):
            raise ValueError("agent requires exactly one of source or file")
        self.task = task
        self.source = source
        self.source_name = source_name or "inline.forma"
        self.file = file
        self.runtime = FormaRuntime(model_provider=provider, tools=tools)

    def run(self, input: dict[str, FormaValue]) -> FormaResult:
        if self.file is not None:
            return self.runtime.run_file(self.file, self.task, input=input)
        return self.runtime.run_task(
            self.source or "",
            self.task,
            input=input,
            source_name=self.source_name,
        )


def agent(
    *,
    task: str,
    provider: ModelProvider | None = None,
    tools: dict[str, object] | None = None,
    source: str | None = None,
    source_name: str | None = None,
    file: str | Path | None = None,
) -> FormaAgent:
    return FormaAgent(
        task=task,
        provider=provider,
        tools=tools,
        source=source,
        source_name=source_name,
        file=file,
    )


def agent_from_package_lock(
    *,
    lock_file: str | Path,
    task: str,
    provider: ModelProvider | None = None,
    tools: dict[str, object] | None = None,
) -> FormaAgent:
    lock_path = Path(lock_file)
    lock = json.loads(lock_path.read_text(encoding="utf8"))
    if lock.get("formaPackageLock") != 1:
        raise ValueError("package lock must have formaPackageLock: 1")

    locked_task = _locked_task(lock, task)
    source_path = lock_path.parent / locked_task["source"]
    source_sha256 = hashlib.sha256(source_path.read_bytes()).hexdigest()
    if source_sha256 != locked_task["sourceSha256"]:
        raise ValueError(f"task source does not match reviewed package lock: {source_path}")
    _verify_package_lock_artifacts(lock, lock_path.parent)

    return agent(
        file=source_path,
        task=task,
        provider=provider or _provider_from_package_lock(lock, lock_path.parent),
        tools=tools,
    )


def _locked_task(lock: dict[str, Any], name: str) -> dict[str, str]:
    for task in lock.get("tasks", []):
        if task.get("name") == name and task.get("source") and task.get("sourceSha256"):
            return task
    raise ValueError(f"{name} is not pinned by the Forma package lock")


def _verify_package_lock_artifacts(lock: dict[str, Any], lock_dir: Path) -> None:
    provider_profile = lock.get("providerProfile")
    if isinstance(provider_profile, dict) and provider_profile.get("path") and provider_profile.get("sha256"):
        _verify_package_lock_hash(lock_dir / provider_profile["path"], provider_profile["sha256"], "provider profile")
    for binding in lock.get("bindings", []):
        if binding.get("output") and binding.get("sha256"):
            _verify_package_lock_hash(lock_dir / binding["output"], binding["sha256"], "generated binding")


def _verify_package_lock_hash(path: Path, expected_sha256: str, artifact_name: str) -> None:
    actual_sha256 = hashlib.sha256(path.read_bytes()).hexdigest()
    if actual_sha256 != expected_sha256:
        raise ValueError(f"{artifact_name} does not match reviewed package lock: {path}")


def _provider_from_package_lock(lock: dict[str, Any], lock_dir: Path) -> ModelProvider:
    provider_profile = lock.get("providerProfile")
    if not isinstance(provider_profile, dict) or not provider_profile.get("path"):
        raise ValueError("package lock providerProfile.path is required when provider is not supplied")
    return provider_from_profile(provider_profile_from_file(lock_dir / provider_profile["path"]))

File: packages/forma-typescript/src/agent.ts
import { createHash } from "node:crypto";
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { FormaRuntime, type ToolHost } from "./runtime.js";
import { providerFromProfile, providerProfileFromFile, type ModelProvider } from "./provider.js";
import type { FormaResult, FormaValue } from "./types.js";

export interface FormaAgent {
  run(input: Record<string, FormaValue>): Promise<FormaResult>;
}

interface FormaAgentBaseOptions {
  task: string;
  provider?: ModelProvider;
  tools?: ToolHost;
}

export type FormaAgentOptions =
  | (FormaAgentBaseOptions & {
      source: string;
      sourceName?: string;
      file?: never;
    })
  | (FormaAgentBaseOptions & {
      file: string;
      source?: never;
      sourceName?: never;
    });

export interface FormaPackageLockAgentOptions {
  lockFile: string;
  task: string;
  provider?: ModelProvider;
  tools?: ToolHost;
}

interface FormaPackageLock {
  formaPackageLock?: number;
  tasks?: Array<{
    name?: string;
    source?: string;
    sourceSha256?: string;
  }>;
  providerProfile?: {
    path?: string;
    sha256?: string;
  };
  bindings?: Array<{
    output?: string;
    sha256?: string;
  }>;
}

export function agent(options: FormaAgentOptions): FormaAgent {
  const runtimeOptions: { modelProvider?: ModelProvider; tools?: ToolHost } = {};
  if (options.provider !== undefined) runtimeOptions.modelProvider = options.provider;
  if (options.tools !== undefined) runtimeOptions.tools = options.tools;
  const runtime = new FormaRuntime(runtimeOptions);

  if (options.file !== undefined) {
    const file = options.file;
    return {
      run(input) {
        return runtime.runFile(file, options.task, { input });
      },
    };
  }

  const source = options.source;
  const sourceName = options.sourceName ?? "inline.forma";
  return {
    run(input) {
      return runtime.runTask(source, options.task, {
        input,
        sourceName,
      });
    },
  };
}

export function agentFromPackageLock(options: FormaPackageLockAgentOptions): FormaAgent {
  const lock = JSON.parse(readFileSync(options.lockFile, "utf8")) as FormaPackageLock;
  if (lock.formaPackageLock !== 1) {
    throw new Error("package lock must have formaPackageLock: 1");
  }
  const task = lock.tasks?.find((candidate) => candidate.name === options.task);
  if (!task?.source || !task.sourceSha256) {
    throw new Error(`${options.task} is not pinned by the Forma package lock`);
  }

  const lockDir = dirname(options.lockFile);
  const sourcePath = join(lockDir, task.source);
  const sourceSha256 = createHash("sha256").update(readFileSync(sourcePath)).digest("hex");
  if (sourceSha256 !== task.sourceSha256) {
    throw new Error(`task source does not match reviewed package lock: ${sourcePath}`);
  }
  verifyPackageLockArtifacts(lock, lockDir);

  const provider = options.provider ?? providerFromPackageLock(lock, lockDir);
  return agent({
    file: sourcePath,
    task: options.task,
    provider,
    ...(options.tools ? { tools: options.tools } : {}),
  });
}

function verifyPackageLockArtifacts(lock: FormaPackageLock, lockDir: string): void {
  const profile = lock.providerProfile;
  if (profile?.path && profile.sha256) {
    verifyPackageLockHash(join(lockDir, profile.path), profile.sha256, "provider profile");
  }
  for (const binding of lock.bindings ?? []) {
    if (binding.output && binding.sha256) {
      verifyPackageLockHash(join(lockDir, binding.output), binding.sha256, "generated binding");
    }
  }
}

function verifyPackageLockHash(path: string, expectedSha256: string, artifactName: string): void {
  const actualSha256 = createHash("sha256").update(readFileSync(path)).digest("hex");
  if (actualSha256 !== expectedSha256) {
    throw new Error(`${artifactName} does not match reviewed package lock: ${path}`);
  }
}

function providerFromPackageLock(lock: FormaPackageLock, lockDir: string): ModelProvider {
  const profilePath = lock.providerProfile?.path;
  if (!profilePath) {
    throw new Error("package lock providerProfile.path is required when provider is not supplied");
  }
  return providerFromProfile(providerProfileFromFile(join(lockDir, profilePath)));
}
